domain/src/main/java/com/thoughtworks/go/config/materials/git/GitMaterial.java

```diff
@@ -443,13 +314,136 @@ public GitMaterial withShallowClone(boolean value) {
         return gitMaterial;
     }
 
-    public String branchWithDefault() {
-        return isBlank(branch) ? GitMaterialConfig.DEFAULT_BRANCH : branch;
+    public String effectiveLocalBranch() {
+        return localBranch(isBlank(refSpecOrBranch) ? GitMaterialConfig.DEFAULT_BRANCH : refSpecOrBranch);
+    }
```

Comment:
One of the main bits changed here; it's important to return the local branch name and not the raw `refspec`. The comparison that references this expects this to be a real branch name.